### PR TITLE
*: online update count in table stats.

### PR DESCRIPTION
**Pull request by hanfei1991**
When insert and update happens, we will cache the changed info in every session. After a duration(five minutes now) passes, handle will sweep every cache and merge them. Then dump the delta info to TiKV.
@shenli @coocood @zimulala @lamxTyler PTAL

### CODE REVIEW

File: sessionctx/variable/session.go [hunk 1]
```diff
@@ -289,6 +295,7 @@ type StatementContext struct {
 		foundRows    uint64
 		warnings     []error
 	}
+	UpdateMapper map[int64]*TableDelta
```
Comment (coocood) [inline sessionctx/variable/session.go]: This should belong to TxnContext.

File: statistics/update_test.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	testKit.MustExec("create table t2 (c1 int, c2 int)")
+
+	rowCount1 := 10
+	rowCount2 := 20
+	for i := 0; i < rowCount1; i ++ {
+		testKit.MustExec("insert into t1 values(1, 2)")
+	}
+	for i := 0; i < rowCount2; i ++ {
+		testKit.MustExec("insert into t2 values(1, 2)")
+	}
```
Comment (coocood) [inline statistics/update_test.go]: Add tests for a transaction that do multiple DML statement.

Comment (coocood) [inline statistics/update_test.go]: Add tests for update and delete

File: statistics/update.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/util/sqlexec"
+)
+
+type updateItem struct {
+	delta int64
+	count int64
+}
+
+type updateMapper map[int64]*updateItem
```
Comment (coocood) [inline statistics/update.go]: s/updateMapper/tableDeltaMap

File: statistics/update.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+		sync.Mutex
+		updateMapper
+	}
+	ctx context.Context
+}
+
+func (m *updateManager) newStatsUpdateHandle() *StatsUpdateHandle {
+	m.listHead.Lock()
+	defer m.listHead.Unlock()
+	mapper := &StatsUpdateHandle{
```
Comment (coocood) [inline statistics/update.go]: The variable name `mapper` is inconsistent with its type.

File: sessionctx/variable/session.go [hunk 5]
```diff
@@ -81,6 +81,21 @@ type TransactionContext struct {
 	InfoSchema    interface{}
 	Histroy       interface{}
 	SchemaVersion int64
+	UpdateMapper  map[int64]*TableDelta
```
Comment (hanfei1991) [inline sessionctx/variable/session.go]: If we don't use pointer, how can we update the structure.

File: statistics/update_test.go [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+	for i := 0; i < rowCount1; i++ {
+		testKit.MustExec("insert into t1 values(1, 2)")
+	}
+	for i := 0; i < rowCount1; i++ {
+		testKit.MustExec("delete from t1 limit 1")
+	}
+	for i := 0; i < rowCount2; i++ {
+		testKit.MustExec("update t2 set c2 = c1")
+	}
+	testKit.MustExec("commit")
```
Comment (coocood) [inline statistics/update_test.go]: Add test for rollback transaction.

File: statistics/update.go [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+	_, err = m.ctx.(sqlexec.SQLExecutor).Execute(fmt.Sprintf("update mysql.stats_meta set version = %d, count = count + %d, modify_count = modify_count + %d where table_id = %d", m.ctx.Txn().StartTS(), delta.Delta, delta.Count, id))
+	if err != nil {
+		return errors.Trace(err)
+	}
+	_, err = m.ctx.(sqlexec.SQLExecutor).Execute("commit")
+	return errors.Trace(err)
+}
+
+// NewStatsUpdateHandle creates a new stats update handle. It is called when a session is created first time.
+func (h *Handle) NewStatsUpdateHandle() *SessionStatsCollector {
```
Comment (coocood) [inline statistics/update.go]: The New function name need to be consistent with the type name.

File: statistics/update.go [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+	return h.updateManager.newSessionStatsCollector()
+}
+
+// DelStatsUpdateHandle deletes a stats update handle from updateManager. It is called when a session is closed.
+func (h *Handle) DelStatsUpdateHandle(handle *SessionStatsCollector) {
```
Comment (coocood) [inline statistics/update.go]: ditto

File: statistics/update.go [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+
+// Update will updates the delta and count for one table id.
+func (h *SessionStatsCollector) Update(id int64, delta int64, count int64) {
+	h.Lock()
+	defer h.Unlock()
+	h.mapper.update(id, delta, count)
+}
+
+// updateManager management the delta information in tidb. It collects the delta map from all sessions and write them to tikv.
+type updateManager struct {
```
Comment (coocood) [inline statistics/update.go]: Do we really need `updateManager`, how about implement the methods in `Handle` directly?

File: statistics/update_test.go (hunk 6 above)
Comment (hanfei1991) [inline statistics/update_test.go]: added

**Comment by hanfei1991** [comment]
@coocood PTAL

**Comment by coocood** [comment]
LGTM

File: statistics/update.go [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+func (m tableDeltaMap) merge(handle *SessionStatsCollector) {
+	handle.Lock()
+	defer handle.Unlock()
+	for id, item := range handle.mapper {
+		m.update(id, item.Delta, item.Count)
+	}
+	handle.mapper = make(tableDeltaMap)
+}
+
+// SessionStatsCollector is a list item that holds the update mapper. If you want write or read mapper, you must add the lock.
```
Comment (lamxTyler) [inline statistics/update.go]: `want to`, `hold the lock`.

File: statistics/update.go [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+	"fmt"
+	"sync"
+
+	"github.com/juju/errors"
+	"github.com/ngaut/log"
+	"github.com/pingcap/tidb/sessionctx/variable"
+	"github.com/pingcap/tidb/util/sqlexec"
+)
+
+type tableDeltaMap map[int64]variable.TableDelta
```
Comment (hanfei1991) [inline statistics/update.go]: In file `sessionctx/variable/session.go`, It is a cache in transaction. If the txn is failed, we should abandon it. Else we should merge it with the collector in session.

**Pull request merged**